src/main/scala/com/microsoft/partnercatalyst/fortis/spark/Pipeline.scala

```diff
@@ -0,0 +1,100 @@
+      val locationsExtractorFactory = new LocationsExtractorFactory(featureServiceClient, geofence).buildLookup()
+      val locationFetcher = locationsExtractorFactory.fetch _
+      val keywordExtractor = new KeywordExtractor(List("Ariana"))
+      val imageAnalyzer = new ImageAnalyzer(ImageAnalysisAuth(Settings.oxfordVisionToken), featureServiceClient)
+      val languageDetector = new LanguageDetector(LanguageDetectorAuth(Settings.oxfordLanguageToken))
+      val sentimentDetectorAuth = SentimentDetectorAuth(Settings.oxfordLanguageToken)
+      val supportedLanguages = Set("en", "fr", "de")
+
+      def convertToSchema(original: T): ExtendedFortisEvent[T] = {
+        val message = analyzer.toSchema(original, locationFetcher, imageAnalyzer)
+        ExtendedFortisEvent(message, Analysis())
+      }
+
+      def addLanguage(event: ExtendedFortisEvent[T]): ExtendedFortisEvent[T] = {
+        val language = analyzer.detectLanguage(event.details, languageDetector)
+        event.copy(analysis = Analysis(language = language))
+      }
+
+      def isLanguageSupported(analysis: Analysis): Boolean = {
+        analysis.language match {
+          case None => false
+          case Some(language) => supportedLanguages.contains(language)
+        }
+      }
+
+      def addKeywords(event: ExtendedFortisEvent[T]): ExtendedFortisEvent[T] = {
+        val keywords = analyzer.extractKeywords(event.details, keywordExtractor)
+        event.copy(analysis = event.analysis.copy(keywords = keywords))
+      }
+
+      def hasKeywords(analysis: Analysis): Boolean = {
+        analysis.keywords.nonEmpty
+      }
+
+      def addEntities(event: ExtendedFortisEvent[T]): ExtendedFortisEvent[T] = {
+        val entities = analyzer.extractEntities(event.details, new PeopleRecognizer(modelsProvider, event.analysis.language))
+        event.copy(analysis = event.analysis.copy(entities = entities))
+      }
+
+      def addSentiments(event: ExtendedFortisEvent[T]): ExtendedFortisEvent[T] = {
+        val sentiments = analyzer.detectSentiment(event.details,
+          new SentimentDetector(modelsProvider, event.analysis.language, sentimentDetectorAuth))
+        event.copy(analysis = event.analysis.copy(sentiments = sentiments))
+      }
+
+      def addLocations(event: ExtendedFortisEvent[T]): ExtendedFortisEvent[T] = {
+        val locations = analyzer.extractLocations(event.details,
+          locationsExtractorFactory.create(Some(new PlaceRecognizer(modelsProvider, event.analysis.language))))
+        event.copy(analysis = event.analysis.copy(locations = locations))
+      }
+
+      // Configure analysis pipeline
+      rdd
+        .map(convertToSchema)
+        .map(addLanguage)
+        .filter(item => isLanguageSupported(item.analysis))
+        .map(addKeywords)
+        .filter(item => hasKeywords(item.analysis))
+        .map(item => addLocations(addSentiments(addEntities(item))))
```

Comment:
I believe the DAG actually optimizes by *combining* the maps calls, only. Imagine the simple case of an `RDD[String]`. If you define a series of maps on it, each map call has a data dependency on the previous. The point of Spark is to partition the data across nodes and then bring transformations to the data. I doubt the DAG will create copies of the same element on each node, distribute different closures, and then somehow (how can it?) merge the elements back together.

I can certainly move these into their own map calls for readability (since I'm almost certain Spark will merge them anyway :) ) TBH, I only just realized that this line is what you were concerned about (forgot I'd combined these :p).